xbmc/music/dialogs/GUIDialogSongInfo.cpp

```diff
@@ -49,6 +46,172 @@ using namespace XFILE;
 
 #define CONTROL_LIST              50
 
+#define TIME_TO_BUSY_DIALOG 500
+
+class CGetSongInfo
+{
+private:
+
+  struct CResult
+  {
+    CResult() : m_event(true), m_success(false) {}
+    CEvent m_event;
+    CFileItemPtr m_song;
+    CFileItemList m_list;
+    bool m_success;
+  };
+
+  struct CGetJob
+    : CJob
+  {
+    CGetJob(std::shared_ptr<CFileItem>& item, CResult& result):
+      m_result(result),
+      m_item(item)
+    {}
+
+    bool DoWork() override
+    {
+      // Fetch full song information including art types list
+      m_result.m_song = m_item;      
+      // Fetch tag data from library using filename, or scanning file (if item not already have this loaded)
+      m_result.m_song->LoadMusicTag();
+      // Fetch album and primary song artist data from library as properties
+      // and lyrics by scanning tags from file 
+      MUSIC_INFO::CMusicInfoLoader::LoadAdditionalTagInfo(m_result.m_song.get());
+
+      // Load song art. 
+      // For songs in library this includes related album and artist(s) art, 
+      // otherwise just embedded or cached thumb is fetched.
+      CMusicThumbLoader loader;
+      loader.LoadItem(m_result.m_song.get());
+
+      // For songs in library fill vector of possible art types, with current art when it exists
+      // for display on the art type selection dialog
+      CMusicArtUtils artUtils;
+      artUtils.FillArtTypesList(*m_result.m_song, m_result.m_list);
+
+      // Fetch artist art for non db songs when artist can be found uniquely by name
+      if (!m_result.m_song->IsMusicDb() && m_result.m_song->HasMusicInfoTag() && !m_result.m_song->GetMusicInfoTag()->GetArtist().empty())
+      {
+        CMusicDatabase db;
+        db.Open();
+        int idArtist = db.GetArtistByName(m_result.m_song->GetMusicInfoTag()->GetArtist()[0]);
+        if (idArtist > 0)
+        {
+          std::map<std::string, std::string> art;
+          if (db.GetArtForItem(idArtist, MediaTypeArtist, art))
+            m_result.m_song->SetArt(art);
+        }
+        db.Close();
+      }
+
+      m_result.m_event.Set();
+      m_result.m_success = true;
+
+      return m_result.m_success;
+    }
+
+    CResult& m_result;
+    std::shared_ptr<CFileItem> m_item;
+  };
+
+public:
+
+  CGetSongInfo(std::shared_ptr<CFileItem>& item)
+    : m_result()
+  {
+    m_id = CJobManager::GetInstance().AddJob(new CGetJob(item, m_result), nullptr, CJob::PRIORITY_HIGH);
+  }
+  ~CGetSongInfo()
+  {
+    CJobManager::GetInstance().CancelJob(m_id);
+  }
+
+  CEvent& GetEvent()
+  {
+    return m_result.m_event;
+  }
+
+  bool Wait(unsigned int timeout)
+  {
+    return m_result.m_event.WaitMSec(timeout);
+  }
+
+  bool GetSongInfo(CFileItem& songitem, CFileItemList& list)
+  {
+    /* if it was not finished or failed, return failure */
+    if (!m_result.m_event.WaitMSec(0) || !m_result.m_success)
+    {
+      list.Clear();
+      return false;
+    }
+
+    list.Copy(m_result.m_list);
+
+    songitem = *m_result.m_song;
```

Comment:
Since I pass a shared_ptr to song fileitem into CGetSongInfo (and then CGetJob), the job popuates it with data directly and this line (in fact the `songitem` parameter) is unnecessary.

But @spiff would it be better to avoid using shared_ptr?

Reply:
So have cleaned up a bit, removing the needless repeats of the same shared_ptr value